quantum/action_tapping.c

```diff
@@ -708,6 +734,110 @@ void waiting_buffer_scan_tap(void) {
+    }
+    ac_dprintf("}\n");
+}
+
+static void process_speculative_press(keyrecord_t *record) {
+    if (num_speculative_keys >= SPECULATIVE_KEYS_SIZE) { // Overflow!
+        ac_dprintf("SPECULATIVE KEYS OVERFLOW: IGNORING EVENT\n");
+        return;
+    }
+
+    const uint16_t keycode = get_record_keycode(record, true);
+    if (!IS_QK_MOD_TAP(keycode)) {
+        return;
+    }
+
+    uint8_t mods = QK_MOD_TAP_GET_MODS(keycode);
+    if ((mods & 0x10) != 0) { // Unpack 5-bit mods to 8-bit representation.
+        mods <<= 4;
+    }
+
+    if ((~(get_mods() | get_speculative_mods()) & mods) != 0 && get_speculative_hold(keycode, record)) {
+        add_speculative_mods(mods);
+        speculative_keys[num_speculative_keys] = (speculative_key_t){
+            .key  = record->event.key,
+            .mods = mods,
+        };
+        ++num_speculative_keys;
+
+        ac_dprintf("Speculative Hold: ");
+        debug_speculative_keys();
+    }
+}
+
+void process_speculative_hold(keyrecord_t *record) {
+    if (num_speculative_keys == 0) {
+        return; // Early return when there are no active speculative keys.
+    }
+
+    // Search for the current event.key in speculative_keys.
+    const keypos_t key = record->event.key;
+    uint8_t        i;
+    for (i = 0; i < num_speculative_keys; ++i) {
+        if (KEYEQ(speculative_keys[i].key, key)) {
+            break;
+        }
+    }
+
+    const uint16_t keycode = get_record_keycode(record, true);
+    if (IS_QK_MOD_TAP(keycode) && record->tap.count == 0) { // MT hold press.
+        if (i < num_speculative_keys) {
+            const uint8_t cleared_mods = speculative_keys[i].mods;
+            del_speculative_mods(cleared_mods);
+            // Don't call send_keyboard_report() here; allow default handling
+            // to reapply the mod before the next report.
+
+            // Remove the ith entry from speculative_keys.
+            --num_speculative_keys;
+            for (uint8_t j = i; j < num_speculative_keys; ++j) {
+                speculative_keys[j] = speculative_keys[j + 1];
+            }
+
+            ac_dprintf("Speculative Hold: settled %02x, ", cleared_mods);
+            debug_speculative_keys();
+        }
+    } else { // Tap press event; cancel speculatively-held mod.
+        if (i >= num_speculative_keys) {
+            i = 0;
+        }
+
+        // Clear mods for the ith key and all keys that follow.
+        uint8_t cleared_mods = 0;
+        for (uint8_t j = i; j < num_speculative_keys; ++j) {
+            cleared_mods |= speculative_keys[j].mods;
+        }
+
+        num_speculative_keys = i; // Remove ith and following entries.
+
+        if ((prev_speculative_mods & cleared_mods) != 0) {
+#        ifdef DUMMY_MOD_NEUTRALIZER_KEYCODE
+            neutralize_flashing_modifiers(get_mods() | prev_speculative_mods);
+#        endif // DUMMY_MOD_NEUTRALIZER_KEYCODE
+        }
+
+        del_speculative_mods(cleared_mods);
+        send_keyboard_report();
```

Comment:
I think we need to sleep for `TAP_CODE_DELAY` here.

Reply:
`neutralize_flashing_modifiers()` calls `tap_code()`, which includes a delay of `TAP_CODE_DELAY` between registering and unregistering the dummy neutralizing keycode. Should that be enough? Or is there another point here where a delay would help?

Reply:
This is the block for `Tap press event; cancel speculatively-held mod.` This function was called from `process_action`, which after we return will go through the usual `process_record_quantum` and such handling to press the key tap. If that is correct, we need to sleep after releasing the mod and sending a keyboard report so that those with the issues that `TAP_CODE_DELAY` is made to mitigate don't experience the speculative mod being applied to the tap.

Reply:
Maybe `TAP_CODE_DELAY` isn't intended for that case. . . but there isn't a better define AFAIK and I believe it was being used between every event of "shift down, tap key down, tap key up, shift up" for Auto Shift before I rewrote it.